### PR TITLE
[New3D] URDF support

### PR DESCRIPTION
**Pull request by jhurliman**
**User-Facing Changes**

- `3D (Beta)` panel renders URDFs from `/robot_description` topic or parameter
- `3D (Beta)` panel supports adding custom URDF layers

**Description**

- Parameter watching support in New3D, subscribe by topic name
- Fix a bug where only one instance of each model was rendered
- Fix custom layers disappearing when adding a new layer
- Mesh loading fixes, report any errors
- Urdfs SceneExtension

Fixes [#3213](https://github.com/foxglove/studio/issues/3213)

Screen Shot 2022-07-13 at 3 48 37 PM
Screen Shot 2022-07-13 at 3 41 10 PM

### CODE REVIEW

**Comment by defunctzombie** [comment]
@jhurliman this does not fix the referenced "no indication my mesh failed to load":

image

This screenshot shows the mesh resource failing to load (the url is not accessible) yet I have no indication in the UI that the mesh failed to load.

**Review by jtbandes — APPROVED** [review]

File: packages/studio-base/src/panels/ThreeDeeRender/ModelCache.ts [hunk 1]
```diff
@@ -54,95 +61,143 @@ export class ModelCache {
...
+async function loadGltf(url: string, reportError: ErrorCallback): Promise<LoadedModel> {
+  const onError = (assetUrl: string) => {
+    const originalUrl = unrewriteUrl(assetUrl);
+    log.error(`Failed to load GLTF asset "${originalUrl}" for "${url}"`);
+    reportError(new Error(`Failed to load GLTF asset "${originalUrl}"`));
+  };
 
-function createDracoLoader(): DRACOLoader {
-  const dracoLoader = new DRACOLoader();
+  const manager = new THREE.LoadingManager(undefined, undefined, onError);
```
[the remark this answers is not in the document]
Comment (jhurliman) [reply packages/studio-base/src/panels/ThreeDeeRender/ModelCache.ts:124]: Yes, this is for reporting a stream of potentially multiple errors

File: packages/studio-base/src/panels/ThreeDeeRender/ModelCache.ts [hunk 2]
```diff
@@ -190,8 +260,56 @@ function fixMaterials(model: LoadedModel): LoadedModel {
...
+      return await (await fetch(dracoDecoderWasmUrl)).arrayBuffer();
+    } else {
+      throw new Error(`DRACOLoader attempt to load non-bundled asset: ${url} as ${responseType}`);
+    }
+  };
+
+  return dracoLoader;
+}
+
+// The THREE.TextureLoader does not support loading .tiff files into textures. To work around
```
[the remark this answers is not in the document]
Comment (jhurliman) [reply packages/studio-base/src/panels/ThreeDeeRender/ModelCache.ts:295]: THREE.TextureLoader is the interface we're directly dealing with that fails on TIFF files. The underlying reason is indeed a lack of browser support (and no manual TIFF decoding in three.js).

File: packages/studio-base/src/panels/ThreeDeeRender/Renderer.ts [hunk 3]
```diff
@@ -386,6 +401,18 @@ export class Renderer extends EventEmitter<RendererEvents> {
     }
   }
 
+  addTopicSubscription<T>(topic: string, handler: (messageEvent: MessageEvent<T>) => void): void {
```
[the remark this answers is not in the document]
Comment (jhurliman) [reply packages/studio-base/src/panels/ThreeDeeRender/Renderer.ts:404]: The relevant change in ThreeDeeRender.tsx is:

```
} else if (topicHandlers.has(topic.name) || datatypeHandlers.has(topic.datatype)) {
```

File: packages/studio-base/src/panels/ThreeDeeRender/renderables/Urdfs.ts [hunk 4]
```diff
@@ -0,0 +1,607 @@
...
+          },
+        });
+      }
+    }
+
+    return entries;
+  }
+
+  override removeAllRenderables(): void {
+    // no-op
```
[the remark this answers is not in the document]
Comment (jhurliman) [reply packages/studio-base/src/panels/ThreeDeeRender/renderables/Urdfs.ts:188]: Seeking shouldn't affect URDFs. When the entire panel is being reset I assume this whole class gets thrown away and reinstantiated.

**Pull request merged**